### PR TITLE
Filter for discarded providers in public API

### PR DESCRIPTION
**Pull request by inulty-dfe**
### Trello
https://trello.com/c/nR0xM6aq/266-delete-providers-that-have-no-region-code-on-manage-and-dont-exist-on-publish

### Context
Apply/Manage needs a way to identify `Provider`s that have been discarded. The current API filters out discarded providers at the definition of the association between `RecruitmentCycle` and `Provider`. We can `unscope` this exclusion in the API for the purposes of returning discarded providers.

This is a non-breaking change to the API. API consumers do not need to be made aware of this change. It can be an undocumented feature if that is preferred.  There is probably no reason a consumer of the API would need to access discarded `Provider`s

### Changes proposed in this pull request

Provider API accepts `filter[discarded]=true` which returns discarded Providers.
When the `filter[discarded]=true` is passed, only providers that are discarded are returned. The attributes include `discarded_at`
Without the filter, `discarded_at` is not returned, keeping the current response consistent.

### Guidance to review
Is this desirable to be added to the Public API?
Could this be made available via another way?
Should this new filter be explicitly documented in public schema?
How should this filter be documented internally?

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [ ] Rebased main
- [x] Cleaned commit history
- [x] Tested by running locally
- [x] Inform data insights team due to database changes

### CODE REVIEW

**Review by tomas-stefano — APPROVED** [review]
LGTM (after checking the documentation part)

**Comment by inulty-dfe** [comment]
Thanks Avin for launching the review app :taco: 

```shell
curl -H 'content-type: application/json' https://publish-review-3737.test.teacherservices.cloud/api/public/v1/recruitment_cycles/2023/providers?filter%5bdiscarded%5d=true | jq
```

File: spec/controllers/api/public/v1/providers_controller_spec.rb
Comment (avinhurry) [inline spec/controllers/api/public/v1/providers_controller_spec.rb:427]: Tiny nit but can we keep it consistent? `param` vs `filter` on line 418?

**Review by avinhurry — COMMENTED** [review]
I’ve tested locally and it works well. Thanks for adding a link in the comments too, very helpful. 

We historically documented all changes made in the API and included them in the release notes. I don’t see any reason to have it as a ‘hidden feature’ but it’s up to @JR-G now. It can be a bit fiddly to work with.

- https://github.com/DFE-Digital/publish-teacher-training/blob/main/docs/source/release-notes.html.md.erb

- https://github.com/DFE-Digital/publish-teacher-training/blob/main/swagger/public_v1/api_spec.json

**Comment by avinhurry** [comment]
> @avinhurry I've updated the OpenAPI spec with the `discarded` filter and the `discarded_at` attributes.

Looks good! @inulty-dfe, should we update the release notes too? https://github.com/DFE-Digital/publish-teacher-training/blob/main/docs/source/release-notes.html.md.erb

**Review by avinhurry — APPROVED** [review]

**Pull request merged**
